### ISSUE TITLE
[QUESTION]: Oversampling / Undersamlpling for imbalance classification

### ISSUE THREAD

**Issue by spyda90** [chemprop/chemprop#572]
Hi,

I'd like to build a categorical predictive models using a binary classification (0 ,1).
The dataset is composed by about 300K data points, but the classification is imbalanced (289K elements with 0 value, 11K with 1 value).

Knowing that such data split can overestimate accuracy, I used f1 as main metrics. The first attempts led to f1-score of about 0.62 in the validation set (using default 80:10:10 split) and a f1-score of 0.33 using an external validation set of 75 data.

Now I'd like to perform oversampling or undersampling in order to improve the performance of the model. My question is: are there some features in Chemprop that can automatically perform a cross-validation using random oversampling / undersampling so the ratio between majority and minority class will be smaller (not imbalanced)?

I am thinking to something like the one reported here: https://machinelearningmastery.com/random-oversampling-and-undersampling-for-imbalanced-classification/ .

Thank you in advance for your help.

Best,
Francesco

**Comment by kevingreenman** [comment]
chemprop doesn't have any built-in features for handling imbalanced data. I recommend looking at something like https://github.com/scikit-learn-contrib/imbalanced-learn to help split your data before using chemprop to train your models.

**Comment by spyda90** [comment]
Hi Kevin,

Thank you for your answer. I guess that what you suggest is the only way to go.

Let's suppose that using scikit-learn I will have my (e.g.) N=10 folded datasets where there will be contain for each one all the 11K "actives" (minority) and 29K "not active" (majority), where the latter will be the results of a random undersampling starting from the original 290K majority class (each cpd not repeated in each fold).

How can I use such (separate) datasets in chemprop then for a single training calculation?

Thank you in advance!

Francesco

**Comment by kevingreenman** [comment]
I apologize - there actually is a (not very well documented) option in chemprop for handling imbalanced classification data: the `--class_balance` flag. However, this is the only option and doesn't give you much flexibility, which is why using external splitting might be more desirable in this case.

There are some [options](https://github.com/chemprop/chemprop/blob/9bc0d0ef483bd6e43ab097bbb5b93a7b065f1fa2/chemprop/args.py#L282-L293) in chemprop for specifying the indicies for your cross-validation sets (particularly the `--index_predetermined` and `--crossval_index_file` arguments). You can look in the [args.py](https://github.com/chemprop/chemprop/blob/master/chemprop/args.py) file to see how these arguments work. But since it's not well documented what format these expect, you may find it easier to just manually do the cross validation by running 10 separate training jobs, each specifying their own `--data_path`, `--separate_val_path`, and `--separate_test_path` (obtained from external splitting code), and then compute the cross validation metrics yourself.

**Comment by kevingreenman** [comment]
Resources for handling imbalanced data:
- https://github.com/ZhiningLiu1998/awesome-imbalanced-learning
- https://github.com/scikit-learn-contrib/imbalanced-learn
- https://github.com/ufoym/imbalanced-dataset-sampler
- https://github.com/YyzHarry/imbalanced-regression

**Comment by spyda90** [comment]
Hi Kevin,

Thank you for all the information.

What I have now are 10 files that acts as training+set+val. From args.py I cannot understand wether --index_predetermined and --crossval_index_file are used to specify the file already splitted in (10) folds or if they are expecting that I have already splitted training and validation sets.

If the latter is the case, I guess I will go for the second suggestion you gave me (10 indipendent trainings).

Francesco

**Comment by kevingreenman** [comment]
`crossval_index_file` should be a pickle file that contains the indices of the datasets you've already split. If you have your splits done outside of chemprop into 10 CSV files of SMILES, then I suggest the independent training approach for now. I'll reply here if I can find an example of creating/using the index file.

**Comment by spyda90** [comment]
That sounds great!
I will update here regarding the performance.

Thank you Kevin

**Comment by kevingreenman** [comment]
note: the question of `crossval_index_file` format was also asked in #265 and was not answered there yet either

**Comment by spyda90** [comment]
Thank you. I took a look at that.
I am still running hyperopt on my undersmapled files. I will keep you update about the results.

Francesco

**Comment by spyda90** [comment]
UPDATE: performing undersampling in order to have 10 dataset with all actives and random 1/10 of inactives (reducing the actives:inactives ratio from 1:25 to 1:3) the performance clearly improved.
I obtained an average f1=0.777 (0.627 before) using internal validation set and f1=0.458 (0.375 before) using an external indipendent validation set of about 70 cpds. The f1=score have been calculated using the average+/-SD of the ten models.
What I noticed is an increased recall performance and a reduction in precision.

I did that using RDKit "D featurization. I guess I should try also without it, since I read that for big size dataset featurization may create noise over D-MPNN.

What do you think? What can I try further?

**Comment by kevingreenman** [comment]
> `crossval_index_file` should be a pickle file that contains the indices of the datasets you've already split. If you have your splits done outside of chemprop into 10 CSV files of SMILES, then I suggest the independent training approach for now. I'll reply here if I can find an example of creating/using the index file.

Posting this for posterity who come searching for help with the proper format for `--crossval_index_file`:

There are some old scripts in the `scripts/` directory of the repo that I think answer this question for us: 
* https://github.com/chemprop/chemprop/blob/master/scripts/create_crossval_splits.py
* https://github.com/chemprop/chemprop/blob/master/scripts/create_crossval_index_files.py

Also, I [used this argument](https://github.com/chemprop/chemprop/issues/310#issuecomment-1197543196) while debugging a previous issue (#310). Using `--save_smiles_splits` while `--num_folds` is greater than 1 should result in the `split_indices.pckl` file being written. But as described in #310, using that `.pckl` file in a subsequent step results in an error.
